File: exercise/23/solution2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        // a fake head
        ListNode head(0);
        ListNode *tail = &head;
        priority_queue<pair<int, ListNode*>, vector<pair<int, ListNode*>>, greater<pair<int, ListNode*>>> q;
        for (int i = 0; i < lists.size(); i++) {
            if (lists[i]) {
                q.push(make_pair(lists[i]->val, lists[i]));
            }
        }
        while (!q.empty()) {
            ListNode *first = q.top().second;
            tail->next = first;
            q.pop();
            if (first->next) {
                q.push(make_pair(first->next->val, first->next));
            }
        }
        return head.next;
    }
};

File: exercise/23/solution2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        from heapq import heappush, heappop
        head = ListNode(0)
        tail = head
        heap = []
        for i in range(len(lists)):
            if lists[i]:
                heappush(heap, (lists[i].val, i))
        while len(heap) > 0:
            no = heap[0][1]
            heappop(heap)
            tail.next = lists[no]
            lists[no] = lists[no].next
            if lists[no]:
                heappush(heap, (lists[no].val, no))
            tail = tail.next
        return head.next
        
